Finish query to pass all current tests

shandyk18 opened the pull request:
The basic algorithm was created for the meeting time query. 

Events and attendees were checked for validity. Afterwards, the events were merged into more complete time ranges for overlapping/nesting events. Potential valid times are found based on the merged invalid events.

shandyk18 commented:
Optional attendees can now be added to queries.

In FindMeetingQuery.java, we keep track of mandatory attendees and their events and all attendees and their events. If all events can produce at least one meeting, that is returned. Otherwise, only mandatory attendees/events are looked at.

wujingya commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+
+    // Add to invalid times collection
+    for (Event event : events) {
+        for(String attendee : attendees) {
```
To make this more efficient, you can use a hashset for attendees, something like:
```
Set<String> attendees = new HashSet<>(request.getAttendees());
for (Event event : events) {
  for (String attendee : event.getAttendees()) {
    if (attendees.contains(attendee)) {
      invalidTimes.add(event.getWhen());
    }
  }
}
```

wujingya commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+    for (int i = 0; i < invalidTimes.size(); i++) {
+        TimeRange invalid = invalidTimes.get(i);
+
+        // if first event at beginning of day, move start to end of event
+        if (invalid.start() == TimeRange.START_OF_DAY && first) {
+            start = invalid.end();
+            first = false;
+            continue;
+        }
```
Seems to me that you don't actually need this.

shandyk18 replied:
You're right! I had this originally because it's leftover from an earlier implementation I did.

wujingya commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -15,9 +15,113 @@
+        return validTimes;
+    }
+
+    invalidTimes = mergeOverlap(invalidTimes);
```
I might be completely stupid, seems to me that we don't need to merge overlap, and findValidTimes should also work just fine?

shandyk18 replied:
When I get rid of the merge, most of my optional attendee tests don't pass and the nestedEvents test doesn't pass. The issue is with nested events and the last if check. If the last event is nested and shorter than the event before it, start will equate itself with the shorter event's end, giving and incorrect range. If I move start to the bottom, then it'll break other conditions.

Day:       |-------------|
Events:       |-------|
                     |---|
Meeting: |---|
                          |---| <-- wrong range

shandyk18 replied:
Added a new variable to account for this. No more merge method!

wujingya replied:
hmmm I see, you're right that wouldn't work. It might work if you keep a latestEndTime and don't add any valid time range if the current invalidTime ends before your latestEndTime. Merging also works so it's up to you.

shandyk18 commented on walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:
```diff
@@ -30,4 +30,14 @@ public void testGreeting() {
+  public void testGreetingTrimsWhitespace() {
+    Greeter greeter = new Greeter();
+
+    String greeting = greeter.greet("Ada");
```
Whoops I remember changing this to have spaces. Not sure why it didn't save? Will change now.

shandyk18 commented:
You're right I didn't really need the merge method. I had to move a few things over from it into the main query function (like sorting) and added a new variable in findValidTimes() to keep track of nesting cases. I also changed attendees and allAttendees into HashSets.
